cloud/scope/azureippool.go

```diff
@@ -0,0 +1,355 @@
+						foundAllocatedIPPool = true
+						allocatedIPPool, foundIPInAllocatedPool = RemoveIPFromAllocatedIPPool(allocatedIPPool, ip)
+						foundIP = foundIP || foundIPInAllocatedPool
+						if err!=nil {
+							return errors.Wrapf(err, "Failed to remove from allocated IP Pool")
+						}
+						azureIPPool.Status.AllocatedIPPools[index] = allocatedIPPool
+					} 
+				}
+				if !foundAllocatedIPPool {
+					return errors.New(fmt.Sprintf("Unable to find allocated IP pool for subnet name %s",subnetName)) 
+				}
+				if !foundIP {
+					return errors.New(fmt.Sprintf("Unable to find IP %s in allocated IP pools",ip))
+				}
+				machineScope.Info(fmt.Sprintf("Allocated IP pools is %v after removing IP %s", azureIPPool.Status.AllocatedIPPools, ip))
+			}
+		}
+
+		machineScope.Info(fmt.Sprintf("Allocated IP Pool is %v", azureIPPool.Status.AllocatedIPPools))
+		machineScope.Info(fmt.Sprintf("Azure IP Pool is %v", azureIPPool))
+		
+		//patch AzureIPPool
+		if err := s.patchHelper.Patch(ctx, azureIPPool); err != nil {
+			return errors.Wrapf(err, "Failed to update Azure IP pool")
+		}
+		machineScope.Info(fmt.Sprintf("Successfully updated AzureIPPool %s", azureIPPool.Name))
+
+		//resetting nic info on machine status
+		machineScope.Info(fmt.Sprintf("Assigning network interfaces to nil"))
+		machineScope.AzureMachine.Status.NetworkInterfaces = []infrav1.NetworkInterface{}
+
+		//resetting reconciledIP info on machine status
+		machineScope.Info(fmt.Sprintf("Setting ReconciledIP to false"))
+		machineScope.AzureMachine.Status.ReconciledIP = false
+
+		//patch azure machine
+		if err := machineScope.patchHelper.Patch(ctx, machineScope.AzureMachine); err != nil {
+			return errors.Wrapf(err, "Failed to update Azure Machine")
+		}
+		machineScope.Info(fmt.Sprintf("Successfully updated AzureMachine status"))	
+		
+	}				
+
+	return nil
+}
+
+func BuildNetworkInterfaceSpec(machineScope *MachineScope,ippool infrav1.IPPool, name string) infrav1.NetworkInterface {
+    networkInterface := infrav1.NetworkInterface{}
+    networkInterface.Name = azure.GenerateNICName(machineScope.Name())
+    networkInterface.VnetName = ippool.VnetName
+    networkInterface.SubnetName = ippool.SubnetName
+    networkInterface.VnetResourceGroup = "aserg"
+    if name == "mgmt-nic-ip-pool" {
+        networkInterface.IsPrimary = true
+    } else {
+		networkInterface.AcceleratedNetworking = true
+    }
+    return networkInterface
+}
+
+func GetPrimaryNetworkInterface(nics []infrav1.NetworkInterface) (infrav1.NetworkInterface, error) {
+    nic := infrav1.NetworkInterface{}
+	for _,nic := range(nics) {
+        if nic.IsPrimary == true {
+            return nic, nil
+        }
+    }
+    return nic, errors.New("Unable to find a nic with isPrimary set to true")
+}
+
+func (s *AzureIPPoolScope) AddToAllocatedIPPool(azureippool *infrav1.AzureIPPool, name string, allocatedIP string, vnetName string, subnetName string) *infrav1.AzureIPPool {
```

Comment:
you don't need the azureippool *infrav1.AzureIPPool parameter.  
You can use s.AzureIPPool instead.

Reply:
The azureippool scope is initialized at the start of every machine reconcile cycle. But within the reconcile IPs function we get an IP from the free IP pool for every ip pool and update the azureippool status each time. That is why the azureippool object in the scope is stale at this time and we have to use the local instance of azureippool that has the status field updated with allocated IPs.

Reply:
i see.